### PR TITLE
Update Gradle config to share common configuration

### PR DESCRIPTION
**Pull request by aSemy**
Hi, I thought I'd make a PR directly rather than an issue - I hope that's okay.

I noticed the Gradle build files had some repetition that could be tidied up by using buildSrc convention plugins. 

I also tidied up some of the Gradle task definitions. 

To help me work with the existing style I added a simple Editor Config file.

If accepted, here are some notes on how to develop in the future:

* The versions of Gradle plugins are defined in one place: `./buildSrc/build.gradle.kts`
* To add a new Gradle plugin, add the Maven coordinates to `./buildSrc/build.gradle.kts`
* Repositories are defined in one place: `./buildSrc/repositories.settings.gradle.kts`
* To change a Kotlin setting for all subprojects, change it in the `kotlin-jvm.gradle.kts` buildSrc plugin. To change a setting for one subproject, change it in that subproject

The versions of dependencies could be defined in one place using [the 
'Java Platform' plugin](https://docs.gradle.org/current/userguide/java_platform_plugin.html) and/or using a [version catalog](https://docs.gradle.org/current/userguide/platforms.html#sub:central-declaration-of-dependencies) - I can implement this if you'd like!

### CODE REVIEW

File: buildSrc/repositories.settings.gradle.kts [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+      jitpack()
+      gradlePluginPortal()
+      mavenCentral()
+    }
+  }
+}
+
+fun RepositoryHandler.jitpack() {
+  maven("https://jitpack.io")
+}
```
Comment (aSemy) [inline buildSrc/repositories.settings.gradle.kts:20]: I added the JitPack repo as a bit of an example. It's not needed, so it can be removed if you prefer.

File: wrapper-generator/build.gradle.kts [hunk 2]
```diff
@@ -21,29 +11,24 @@ dependencies {
...
+    testImplementation(projects.library)
 }
 
 application {
     mainClass.set("it.krzeminski.githubactions.wrappergenerator.GenerationEntryPointKt")
-    tasks.run.get().workingDir = rootProject.projectDir
 }
 
-tasks.getByName("run") {
-    finalizedBy(":library:ktlintFormat")
```
Comment (aSemy) [inline wrapper-generator/build.gradle.kts:37]: I wasn't sure what the purpose of this link was. Normally projects aren't supposed to reference other projects. 

> "Don’t reference other project tasks directly"
> 
> https://docs.gradle.org/current/userguide/cross_project_publications.html

If it's necessary, maybe there's an idiomatic way. What's the intention for this link?

Comment (jmfayard) [reply wrapper-generator/build.gradle.kts:37]: `wrapper-generator` produces the Actions Wrappers from https://krzema12.github.io/github-actions-kotlin-dsl/supported-actions/ inside the `library` module. After producing the you see all kind of changes but they are only noises, running`ktlintFormat` is necessary to eliminate them.

**Comment by jmfayard** [comment]
For dependencies, [refreshVersions](https://github.com/jmfayard/refreshVersions) can take care of both generating the versions catalog and finding available updates instead of dependabot, but I don't know what @krzema12 thinks

File: wrapper-generator/build.gradle.kts (hunk 2 above)
Comment (aSemy) [reply wrapper-generator/build.gradle.kts:37]: Okay, thanks. I think I get it now. This is a bit tricky to untangle because it's a circular dependency. `library` depends on `wrapper-generator`, but `wrapper-generator` tests depends on `library`...

For now I've added back in the link - if it works, it works!

I have some thoughts on doing it idiomatically, but this could be a fair bit of work. 

`wrapper-generator` is basically a bunch of Gradle tasks, but it's missing out on the Gradle extras like incremental work and task avoidance. `wrapper-generator` could be converted into a Gradle plugin that's an 'included build', and each task that's in `WrappersToGenerate.kt` would be a Gradle task, defined in some `build.gradle.kts`. 

The tests in `wrapper-generator` might benefit from the Gradle Java 'test fixtures' plugin, and setting up a custom Gradle 'configuration' to share the generated wrappers. This might avoid the 'circular dependency' link, as while each project would depend on the other, they wouldn't depend on each other at the same time - 

1. first `wrapper-generator` would compile and execute, and produce some output files. 
2. `library` would consume those files, and build itself
3. `wrapper-generator` would do a test-fixtures compile, consume the built `library`, and run the tests

I'm not sure if that would work though.

So yeah, for now, referencing another project's task is probably best!

**Comment by krzema12** [comment]
> For dependencies, [refreshVersions](https://github.com/jmfayard/refreshVersions) can take care of both generating the versions catalog and finding available updates instead of dependabot, but I don't know what @krzema12 thinks

@jmfayard to be honest, Dependabot does a pretty good job here. It creates PRs that I can review and merge from my phone, no need to actually change something in the code. As far as I see, refreshVersions takes a different approach by adding comments with available versions, which makes the process slightly more complicated. I know it's your library so please don't take it personally, I'm just pragmatically choosing an option that means less effort for me :)

File: wrapper-generator/build.gradle.kts (hunk 2 above)
Comment (krzema12) [reply wrapper-generator/build.gradle.kts:37]: I appreciate all the improvements and ideas, but let's take it easy and in small steps :) For now this PR is already a nice improvement, extracting repeated pieces of config to local plugins. As I'm not very proficient with Gradle, I'll review it thorougly and I have to find time to do it comfortably, within a week or two.

File: build.gradle.kts [hunk 3]
```diff
@@ -1,8 +1,6 @@
...
-    kotlin("jvm") version "1.6.20" apply false
-    kotlin("plugin.serialization") version "1.6.20" apply false
+group = "it.krzeminski"
+version = "0.13.0"
 
-    // Code quality.
-    id("io.gitlab.arturbosch.detekt") version "1.20.0" apply false
-    id("org.jlleitschuh.gradle.ktlint") version "10.2.1" apply false
+tasks.wrapper {
+    gradleVersion = "7.4.2"
```
Comment (krzema12) [inline build.gradle.kts:5]: Why is it needed? Wrapper's version is already defined in https://github.com/krzema12/github-actions-kotlin-dsl/blob/f618aa339ac6ff3d51b5c0d7d6ffd199d46020dc/gradle/wrapper/gradle-wrapper.properties#L4

File: gradle/wrapper/gradle-wrapper.properties [hunk 4]
```diff
@@ -1,6 +1,5 @@
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
-distributionSha256Sum=29e49b10984e585d8118b7d0bc452f944e386458df27371b49b4ac1dec4b7fda
```
Comment (krzema12) [inline gradle/wrapper/gradle-wrapper.properties:3]: Why is it removed? I remember the lack of it caused some issues in the IDE.

File: build.gradle.kts [hunk 5]
```diff
@@ -1,8 +1,6 @@
-plugins {
-    kotlin("jvm") version "1.6.20" apply false
-    kotlin("plugin.serialization") version "1.6.20" apply false
+group = "it.krzeminski"
+version = "0.13.0"
```
Comment (krzema12) [inline build.gradle.kts:2]: Why is this version in the root project? True that for now only the "library" subproject is published, but there's a chance we'll publish more artifacts, not necessarily with the same version. Is it a problem to have it in library/build.gradle.kts?

File: .editorconfig [hunk 6]
```diff
@@ -0,0 +1,9 @@
+root = true
+
+[{*.kt, *.kts}]
+charset = utf-8
```
Comment (krzema12) [inline .editorconfig:4]: Thanks for adding it :)

File: buildSrc/repositories.settings.gradle.kts [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+    pluginManagement {
+        repositories {
+            jitpack()
+            gradlePluginPortal()
+            mavenCentral()
+        }
+    }
+}
+
+fun RepositoryHandler.jitpack() {
```
Comment (krzema12) [inline buildSrc/repositories.settings.gradle.kts:18]: I guess it's not needed :)

File: settings.gradle.kts [hunk 8]
```diff
@@ -1,11 +1,20 @@
 rootProject.name = "github-actions-kotlin-dsl"
 
+apply(from = "./buildSrc/repositories.settings.gradle.kts")
```
Comment (krzema12) [inline settings.gradle.kts:3]: I'd rather stay with defining repositories per module, to be explicit and have more control over which module uses what. E. g. we're more liberal about wrapper-generator or script-generator modules, while for the library we'd like to avoid using certain repositories.

**Review by krzema12 — CHANGES_REQUESTED** [review]
A lot of small changes, not all of them seem really necessary to me. Anyway, thanks for this - if it doesn't reach the main branch in such form, I'll treat it as general inspiration that certain pieces could be extracted to local convention plugins.

**Comment by krzema12** [comment]
Side-note: due to my chronical lack of time, such refactoring would be much easier for me to review if it tackled a smaller bit at once. E. g. extracting publishing or applying ktlint and detekt to all projects. Otherwise it's hard to follow which change relates to what kind of improvement.

File: build.gradle.kts (hunk 3 above)
Comment (aSemy) [reply build.gradle.kts:5]: The task is used to update the Gradle wrapper version, and `gradle-wrapper.properties` is updated by the task. 

I like defining it in the task so I can update the version in `build.gradle.kts` and run `./gradlew wrapper`, and it will update the URL in `gradle-wrapper.properties`. The alternative is running `./gradlew wrapper --gradle-version 7.4.2`, but I always forget the exact argument :)

File: build.gradle.kts (hunk 5 above)
Comment (aSemy) [reply build.gradle.kts:2]: I set the version in the root and created [a `subproject.build.kts` convention plugin](https://github.com/krzema12/github-actions-kotlin-dsl/blob/74f2feba52fe7162796f147013e404fee6d824d3/buildSrc/src/main/kotlin/buildsrc/convention/subproject.gradle.kts), which is manually applied to all subproject.

Because it's a convention plugin, any subproject can override the convention.

```kotlin
plugins {
  buildsrc.convention.subproject
}

// override the convention supplied by 'buildsrc.convention.subproject'
version = "1.2.3"
```

Personally I like having a unified version, but if you want to have separate versions per subproject you can still do that. At the moment the PR just provides a default (defined in the root project).

File: settings.gradle.kts (hunk 8 above)
Comment (aSemy) [reply settings.gradle.kts:3]: That's possible, repositories defined in the subprojects will override the centralized definitions 

https://docs.gradle.org/current/userguide/declaring_repositories.html#sub:centralized-repository-declaration

> By default, repositories declared by a project will override whatever is declared in settings.

I can't think of a reason for having different repos per subproject though...

**Comment by krzema12** [comment]
FYI I started applying this PR in smalller chunks (some smaller changes already merged), but I'm struggling to have a smooth experience with IntelliJ and these local convention plugins. I faced some issue with multiple Kotlin compiler artifacts where Gradle refused to proceed, it's also not trivial to get what Kotlin versions are safe to use where. It got me thinking if we really want to have these convention plugins. I'll give it another shot later.

**Comment by aSemy** [comment]
> FYI I started applying this PR in smalller chunks (some smaller changes already merged), but I'm struggling to have a smooth experience with IntelliJ and these local convention plugins. I faced some issue with multiple Kotlin compiler artifacts where Gradle refused to proceed, it's also not trivial to get what Kotlin versions are safe to use where. It got me thinking if we really want to have these convention plugins. I'll give it another shot later.

Yes the experience can be really tough. I have lots of trouble because in buildSrc IntelliJ flags lots of code Gradle Kotlin DSL shortcuts as invalid, even though it's not. It makes it hard to figure out what's right and what's wrong.

The Gradle/Kotlin compatibility is documented here: https://docs.gradle.org/current/userguide/compatibility.html#kotlin. Currently Gradle will set the Kotlin language level to 1.4, which it is quite strict about. Gradle uses an embedded version of Kotlin, which is 1.5.x at the moment - but you can safely use 1.6.21 in `./buildSrc/build.gradle.kts` 

In Gradle 7.5 the embedded version of Kotlin will be bumped to 1.6.21, but the language level will **still** be set to 1.4 - it's annoying how Gradle is lagging behind.

**Comment by krzema12** [comment]
https://github.com/dependabot/dependabot-core/issues/2180

Is it true that after this change we won't have Dependabot updates? It's a major disadvantage to me 😢

**Comment by aSemy** [comment]
> [dependabot/dependabot-core#2180](https://github.com/dependabot/dependabot-core/issues/2180)
> 
> Is it true that after this change we won't have Dependabot updates? It's a major disadvantage to me 😢

Hmm maybe. Does Dependabot recommend updates for Kotlin, Detekt, and ktlint Gradle plugins? Because those are the only versions that are defined within `buildSrc`: https://github.com/krzema12/github-actions-kotlin-dsl/pull/194/files#diff-7b5c40eb5522b5216cfe8c75612e579a8db0daee9ded80df9ce6a56e8361a962


The versions of dependencies are still set in the regular subproject `build.gradle.kts` files, like before. But I don't know how Dependabot works with that. I've not used Dependabot before, but from reading the issues it seems like Dependabot has never really fully supported Gradle. [Apparently Renovate does](https://github.com/dependabot/dependabot-core/issues/3121#issuecomment-1078025336)?

Personally I like using version catalogs (all the dependencies and versions are defined in a single file, `libs.versions.toml`), so I only have to check versions in single place, and use a version platform (Gradle `java-platform` plugin) to keep versions (including transitive dependencies) aligned. Hopefully Dependabot supports that soon https://github.com/dependabot/dependabot-core/issues/3121

**Comment by krzema12** [comment]
Great, thanks, I'll give Renovate a shot!

**Comment by krzema12** [comment]
@aSemy would you be so kind to rebase these changes on newest main branch? Most of the minor ones should be already applied, and I'm most interested in using the precompiled plugins. Renovate seems to support these. Regarding version catalog, let's hold with it for now, we may try it in a separate PR. In this one I'd like to focus only on extracting stuff to precompiled plugins.

**Comment by aSemy** [comment]
> @aSemy would you be so kind to rebase these changes on newest main branch? Most of the minor ones should be already applied, and I'm most interested in using the precompiled plugins. Renovate seems to support these. Regarding version catalog, let's hold with it for now, we may try it in a separate PR. In this one I'd like to focus only on extracting stuff to precompiled plugins.

Done!

I refreshed some of the implementations based on some Gradle things I've learned.

I also created a separate task type for that 'check Maven Central' script task, as a bit of a demo to show how to do that.

I didn't test this significantly, so some things might be janky, or maybe I didn't properly refresh the versions after the merge

File: buildSrc/build.gradle.kts [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+object Versions {
+    const val kotlin = "1.7.10"
+    const val detekt = "1.21.0"
+    const val ktlint = "10.3.0"
+    const val nexusPublishPlugin = "1.1.0"
+}
+
+dependencies {
+    implementation(enforcedPlatform("org.jetbrains.kotlin:kotlin-bom:${Versions.kotlin}"))
+    implementation("org.jetbrains.kotlin:kotlin-gradle-plugin:${Versions.kotlin}")
```
Comment (krzema12) [inline buildSrc/build.gradle.kts:15]: I'm afraid Renovate won't be able to detect such dependency string and prepare a PR upgrading it. Could we inline the versions?

File: buildSrc/src/main/kotlin/buildsrc/convention/kotlin-jvm.gradle.kts [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+
+tasks.withType<KotlinCompile> {
+    kotlinOptions {
+        apiVersion = "1.5"
+        languageVersion = "1.7"
+
+        allWarningsAsErrors = true
+
+        freeCompilerArgs += listOf(
+            "-opt-in=kotlin.RequiresOptIn",
```
Comment (krzema12) [inline buildSrc/src/main/kotlin/buildsrc/convention/kotlin-jvm.gradle.kts:33]: It shouldn't be needed anymore, we've removed it recently here: https://github.com/krzema12/github-actions-kotlin-dsl/commit/2c13d9d4fa520ea8132e7020ca54c65740935029

File: library/src/gen/kotlin/it/krzeminski/githubactions/actions/actions/SetupPythonV4.kt [hunk 11]
```diff
@@ -21,8 +21,8 @@ import kotlin.collections.toTypedArray
  */
 public class SetupPythonV4(
     /**
-     * Version range or exact version of Python to use, using SemVer's version range syntax. Reads
-     * from .python-version if unset.
+     * Version range or exact version of Python or PyPy to use, using SemVer's version range syntax.
```
Comment (krzema12) [inline library/src/gen/kotlin/it/krzeminski/githubactions/actions/actions/SetupPythonV4.kt:24]: This is some unrelated change to the wrapper - let's revert it.

**Review by krzema12 — CHANGES_REQUESTED** [review]
Thanks! Several minor things.

File: build.gradle.kts [hunk 12]
```diff
@@ -1,18 +1,21 @@
...
-    id("io.github.gradle-nexus.publish-plugin") version "1.1.0" // Needs to be applied to the root project.
+    id("io.github.gradle-nexus.publish-plugin") // Needs to be applied to the root project.
+
+    // note: plugin versions are defined in ./buildSrc/build.gradle.kts
 }
 
+group = "it.krzeminski"
+version = "0.23.0"
+
 nexusPublishing {
```
Comment (krzema12) [inline build.gradle.kts:12]: Do you know if it's possible to move to the plugin responsible for publishing?

File: buildSrc/src/main/kotlin/buildsrc/config/publishing.kt [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+ *
+ * @param[configure] additional configuration. Can also be used to override the defaults.
+ */
+fun MavenPublication.createGitHubActionsKotlinDslPom(
+    libraryName: String,
+    githubUser: String,
+    configure: MavenPom.() -> Unit = {},
+) = pom {
+    name.set(libraryName)
+    description.set("Authoring GitHub Actions workflows in Kotlin.")
```
Comment (krzema12) [inline buildSrc/src/main/kotlin/buildsrc/config/publishing.kt:17]: If we really wanted to make a reusable function here, this description would need to be parameterized as well. However, I suggest inlining this in the publishing plugin for simplicity since there's a single library being published now, and we can always refactor in case another library is published.

File: library/build.gradle.kts [hunk 14]
```diff
@@ -48,125 +32,35 @@ sourceSets {
...
 
-tasks.withType<Test> {
-    useJUnitPlatform()
-}
-
-java {
-    withJavadocJar()
-    withSourcesJar()
+val waitUntilLibraryPresentInMavenCentral by tasks.registering(AwaitMavenCentralDeployTask::class) {
+    queriedUrl.set("https://repo1.maven.org/maven2/it/krzeminski/github-actions-kotlin-dsl/$version/")
```
Comment (krzema12) [inline library/build.gradle.kts:44]: I like the idea of extracting a task for it :) However we could make its API a bit more convenient: instead of passing the URL, we could pass groupId, artifactId and version (BTW, all of these possibly taken from Gradle, instead of hardcoding), and the URL would be built inside the task.

File: settings.gradle.kts [hunk 15]
```diff
@@ -1,11 +1,20 @@
...
+
 include(
-    "library",
-    "wrapper-generator",
-    "script-generator"
+    ":library",
```
Comment (krzema12) [inline settings.gradle.kts:6]: BTW, I cannot wrap my head around when it's needed to use the colons. Do you have a good source at hand explaining it?

File: library/src/gen/kotlin/it/krzeminski/githubactions/actions/actions/SetupPythonV4.kt (hunk 11 above)
Comment (krzema12) [reply library/src/gen/kotlin/it/krzeminski/githubactions/actions/actions/SetupPythonV4.kt:24]: If it's a wrapper change that needs to be made, don't worry about the workflow that will fail because of that. We have a process that updates it :)

File: build.gradle.kts [hunk 16]
```diff
@@ -1,18 +1,21 @@
...
     repositories {
         sonatype()
     }
     packageGroup.set("it.krzeminski")
 }
+
+tasks.wrapper {
+    gradleVersion = "7.5"
```
Comment (krzema12) [inline build.gradle.kts:20]: Will Renovate update it as well? I personally doubt it, so I'd remove this task. Renovate handles Gradle wrapper updates nicely so I don't even have to think about it anymore.

File: script-generator/build.gradle.kts [hunk 17]
```diff
@@ -25,13 +20,12 @@ dependencies {
...
-tasks.withType<Test> {
-    useJUnitPlatform()
-}
-
 application {
     mainClass.set("it.krzeminski.githubactions.scriptgenerator.MainKt")
-    tasks.run.get().workingDir = rootProject.projectDir
+}
+
+tasks.run.configure {
```
Comment (krzema12) [inline script-generator/build.gradle.kts:27]: Is there a need to change it? I liked that it was in the `applocation` block, fits there well.

File: settings.gradle.kts [hunk 18]
```diff
@@ -1,11 +1,20 @@
...
+    ":wrapper-generator",
+    ":script-generator",
 )
 
+enableFeaturePreview("TYPESAFE_PROJECT_ACCESSORS")
```
Comment (krzema12) [inline settings.gradle.kts:11]: We've already got it at the very bottom of the file.

File: build.gradle.kts (hunk 12 above)
Comment (aSemy) [reply build.gradle.kts:12]: Yup, totally possible - you can add an `if(project == rootProject) { }` check, and apply the Nexus publishing config in the block.

I'm not too keen on it though. The convention plugins should be re-usable config that can be applied to multiple subprojects. If some config is only applicable to a single project, then I think it's fine that the config is just defined in that project's `build.gradle.kts`.

If you want to change it, feel free to make a commit :)

File: library/build.gradle.kts (hunk 14 above)
Comment (aSemy) [reply library/build.gradle.kts:44]: Yup, can do. Just make sure to define the input variables like `@get:Input abstract val groupId` so Gradle will [manage the properties](https://docs.gradle.org/current/userguide/custom_gradle_types.html#managed_properties), and the [task will be skipped if it's up-to-date](https://docs.gradle.org/current/userguide/more_about_tasks.html#sec:up_to_date_checks).

File: settings.gradle.kts (hunk 15 above)
Comment (aSemy) [reply settings.gradle.kts:6]: The first colon is optional, it signifies the root project. I like to include it here to make it clear that these are subprojects and are _not_ directories. It also looks nicer when there are nested subprojects.

https://docs.gradle.org/current/userguide/intro_multi_project_builds.html#sec:executing_tasks_by_fully_qualified_name

Feel free to remove it 👍

File: script-generator/build.gradle.kts (hunk 17 above)
Comment (aSemy) [reply script-generator/build.gradle.kts:27]: Changing `get()` to `configure {}` helps with [Gradle configuration avoidance](https://docs.gradle.org/current/userguide/task_configuration_avoidance.html). `get()` will eagerly create a task, while `configure {}` will only configure the task when it's needed. It's a very minor change in this case - but it's nice to consistently, and the small changes will stack up.

In this case you probably could configure a task inside the `application {}` block, but generally It's not good practice - it's not clearly defined *when* the block will be run. In this case it's almost certainly fine, but I think a consistent approach (configure everything separately) is best. You can change it if you'd like!

File: buildSrc/repositories.settings.gradle.kts (hunk 1 above)
Comment (aSemy) [reply buildSrc/repositories.settings.gradle.kts]: removd

File: buildSrc/src/main/kotlin/buildsrc/config/publishing.kt (hunk 13 above)
Comment (aSemy) [reply buildSrc/src/main/kotlin/buildsrc/config/publishing.kt:17]: I like it because, even if it's not re-used, it keeps the build script smaller and the key config more visible. You can merge it back if you like though :)

File: buildSrc/src/main/kotlin/buildsrc/convention/publishing.gradle.kts [hunk 19]
```diff
@@ -0,0 +1,48 @@
...
+
+import buildsrc.config.createGitHubActionsKotlinDslPom
+
+plugins {
+    `maven-publish`
+    signing
+}
+
+val githubUser = "krzema12"
+val libraryName = "github-actions-kotlin-dsl"
```
Comment (krzema12) [inline buildSrc/src/main/kotlin/buildsrc/convention/publishing.gradle.kts:11]: How can I pass parameters to such convention plugin? Let's say I want to reuse it to publish another Maven artifact.

File: buildSrc/build.gradle.kts [hunk 20]
```diff
@@ -0,0 +1,27 @@
+import org.jetbrains.kotlin.gradle.tasks.KotlinCompile
+
+plugins {
+    `kotlin-dsl`
+    kotlin("jvm") version embeddedKotlinVersion
+}
+
+dependencies {
+    implementation(enforcedPlatform("org.jetbrains.kotlin:kotlin-bom:1.7.10"))
```
Comment (krzema12) [inline buildSrc/build.gradle.kts:9]: Is there any reason to use `enforcedPlatform` here and `platform` several lines below? Do they differ somehow?

File: library/build.gradle.kts [hunk 21]
```diff
@@ -1,31 +1,15 @@
...
 plugins {
-    kotlin("jvm")
+    buildsrc.convention.`kotlin-jvm`
+    buildsrc.convention.publishing
+
     kotlin("plugin.serialization")
 
     // Code quality.
     id("io.gitlab.arturbosch.detekt")
     id("org.jlleitschuh.gradle.ktlint")
```
Comment (krzema12) [inline library/build.gradle.kts:12]: Can be removed right? It's included in the kotlin-jvm plugin.

File: script-generator/build.gradle.kts (hunk 17 above)
Comment (krzema12) [reply script-generator/build.gradle.kts:27]: Thanks for the explanation! I asked mostly to learn if it's directly related to creating convention plugins, or it's a "by the way" change :)

File: buildSrc/build.gradle.kts (hunk 20 above)
Comment (krzema12) [reply buildSrc/build.gradle.kts:9]: > However often BOMs are not only providing versions as recommendations, but also a way to override any other version found in the graph. You can enable this behavior by using the enforcedPlatform keyword, instead of platform, when importing the BOM

https://docs.gradle.org/current/userguide/platforms.html#sub:bom_import

In this very case it probably doesn't make a big difference right? But I'd like to be consistent.

**Comment by krzema12** [comment]
In https://github.com/krzema12/github-actions-kotlin-dsl/pull/386 I'm merging a part of this PR - just the `kotlin-jvm` plugin and centralized repos management, to see how they work in practice and to not touch releasing config for now. If everything goes fine and e.g. Renovate works fine in this configuration, I'll proceed with introducing changes in this PR :smile:

File: buildSrc/src/main/kotlin/buildsrc/convention/publishing.gradle.kts (hunk 19 above)
Comment (aSemy) [reply buildSrc/src/main/kotlin/buildsrc/convention/publishing.gradle.kts:11]: There are two options

#### 1. Extra properties

Simplest and easiest, but *slightly* less typesafe, and doesn't work with auto complete. It's best for this situation.

* https://docs.gradle.org/current/dsl/org.gradle.api.plugins.ExtraPropertiesExtension.html
* https://docs.gradle.org/current/userguide/kotlin_dsl.html#extra_properties

```kotlin
// ./buildSrc/.../my-convention-plugin.gradle.kts

val githubUser: String by extra("default")

println(githubUser) // will always print 'default', because the subproject hasn't overridden it yet

tasks.build {
  doLast {
    println(githubUser) // should print 'abcd' in subproject-a
  }
}
```

```kotlin
// ./subproject-a/build.gradle.kts
plugins {
  `my-convention-plugin`
}

val githubUser: String by extra("abcd")
```

`githubUser` has a default value, `default`. The subproject overrides it.

`githubUser` *looks* like a String property, but under the hood it's a `Provider<String>` https://docs.gradle.org/current/userguide/kotlin_dsl.html#extra_properties

#### 2. a typed extension

Much more typesafe. These are best for publishing a plugin for general useage, but are probably overkill for an internal build plugin

https://docs.gradle.org/current/userguide/custom_plugins.html#sec:getting_input_from_the_build

File: buildSrc/build.gradle.kts (hunk 20 above)
Comment (aSemy) [reply buildSrc/build.gradle.kts:9]: yeah, there's not a big difference here, `platform` will be fine.

File: library/build.gradle.kts (hunk 21 above)
Comment (aSemy) [reply library/build.gradle.kts:12]: ah yes, probably. I think that I left it there by mistake during a merge

**Comment by krzema12** [comment]
Renovate works fine for the precompiled plugins: https://github.com/krzema12/github-actions-kotlin-dsl/commit/1832f83159d5b4896e313610312e342dbba8525a

File: buildSrc/build.gradle.kts (hunk 20 above)
Comment (Goooler) [inline buildSrc/build.gradle.kts:9]: Maybe we should `includeBuild` a module instead of using `buildSrc`, it's slow.

Comment (aSemy) [reply buildSrc/build.gradle.kts:9]: What about it is slow? 

I'm not an expect, but the only drawback I know about is that any changes in buildSrc require a full rebuild. But that happens pretty rarely, and most Gradle projects are small enough, or change buildSrc so rarely, that a full rebuild isn't a heavy hit. But overall buildSrc is much more easy to configure and maintain.

Comment (Goooler) [reply buildSrc/build.gradle.kts:9]: Yeah, that's it.

**Comment by krzema12** [comment]
It took me a while to understand and merge this PR (in several pieces), but we finally did it! :muscle: Thanks a lot for sharing your knowledge about Gradle and answering all my question in detail. I just became a more conscious user of it :wink: Plus the code base got cleaner. I'm open to any future improvements, but please let's make them in smaller, easier to digest pieces. Best if one thing is changed at a time - no confusion :arrow_right: fast merge :smiley:

Cheers!
